File: cpp/base/src/main/vsdk/toolkit/environment/geometry/surface/QuadMesh.cpp
#include "vsdk/toolkit/environment/geometry/surface/QuadMesh.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/environment/geometry/surface/TriangleMesh.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/environment/geometry/surface/TriangleMeshGroup.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/environment/geometry/elements/Vertex.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/environment/geometry/elements/Ray.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/environment/geometry/elements/RayHit.h"
#include "java/lang/String.h"
#include "java/util/ArrayList.txx"
#include "java/lang/String.h"
#include <algorithm>
#include "java/lang/String.h"

QuadMesh::QuadMesh() : name("default")
{
}

java::String QuadMesh::getName() const
{
    return name;
}

void QuadMesh::setName(const java::String& inName)
{
    name = inName;
}

void QuadMesh::getVertexAt(int i, Vertex& vertex) const
{
    vertex.position = Vector3Dd(vertexPositions.get(3*i), vertexPositions.get(3*i+1), vertexPositions.get(3*i+2));
    if ((int)vertexNormals.size() >= (i+1)*3) {
        vertex.normal = Vector3Dd(vertexNormals.get(3*i), vertexNormals.get(3*i+1), vertexNormals.get(3*i+2));
    }
    if ((int)vertexBinormals.size() >= (i+1)*3) {
        vertex.binormal = Vector3Dd(vertexBinormals.get(3*i), vertexBinormals.get(3*i+1), vertexBinormals.get(3*i+2));
    }
    if ((int)vertexTangents.size() >= (i+1)*3) {
        vertex.tangent = Vector3Dd(vertexTangents.get(3*i), vertexTangents.get(3*i+1), vertexTangents.get(3*i+2));
    }
    if ((int)vertexUvs.size() >= (i+1)*2) {
        vertex.u = vertexUvs.get(2*i);
        vertex.v = vertexUvs.get(2*i+1);
    }
}

void QuadMesh::initVertexPositionsArray(int n)
{
    vertexPositions.clear(); vertexPositions.reserve((long int)n*3);
    for (long int i = 0; i < (long int)n*3; i++) vertexPositions.add(0.0);
    incidentQuadsPerVertexArray.clear(); incidentQuadsPerVertexArray.reserve((long int)n);
    for (long int i = 0; i < (long int)n; i++) incidentQuadsPerVertexArray.add(java::ArrayList<int>());
}

void QuadMesh::initVertexColorsArray()
{
    long int nn = (long int)getNumVertices()*3;
    vertexColors.clear(); vertexColors.reserve(nn);
    for (long int i = 0; i < nn; i++) vertexColors.add(0.0);
}

void QuadMesh::initVertexNormalsArray()
{
    long int nn = (long int)getNumVertices()*3;
    vertexNormals.clear(); vertexNormals.reserve(nn);
    for (long int i = 0; i < nn; i++) vertexNormals.add(0.0);
}

void QuadMesh::setVertexes(const java::ArrayList<Vertex>& vertexes)
{
    initVertexPositionsArray((int)vertexes.size());
    initVertexNormalsArray();
    for (int i = 0; i < (int)vertexes.size(); i++) {
        Vertex v = vertexes.get(i);
        vertexPositions[3*i] = v.position.x();
        vertexPositions[3*i+1] = v.position.y();
        vertexPositions[3*i+2] = v.position.z();
        vertexNormals[3*i] = v.normal.x();
        vertexNormals[3*i+1] = v.normal.y();
        vertexNormals[3*i+2] = v.normal.z();
    }
}

void QuadMesh::initQuadArrays(int n)
{
    quadIndices.clear(); quadIndices.reserve((long int)n*4);
    for (long int i = 0; i < (long int)n*4; i++) quadIndices.add(0);
}

void QuadMesh::setVertexAt(int i, const Vertex& vertex)
{
    vertexPositions[3*i] = vertex.position.x();
    vertexPositions[3*i+1] = vertex.position.y();
    vertexPositions[3*i+2] = vertex.position.z();
    if ((int)vertexNormals.size() >= (i+1)*3) {
        vertexNormals[3*i] = vertex.normal.x();
        vertexNormals[3*i+1] = vertex.normal.y();
        vertexNormals[3*i+2] = vertex.normal.z();
    }
    if ((int)vertexBinormals.size() >= (i+1)*3) {
        vertexBinormals[3*i] = vertex.binormal.x();
        vertexBinormals[3*i+1] = vertex.binormal.y();
        vertexBinormals[3*i+2] = vertex.binormal.z();
    }
    if ((int)vertexTangents.size() >= (i+1)*3) {
        vertexTangents[3*i] = vertex.tangent.x();
        vertexTangents[3*i+1] = vertex.tangent.y();
        vertexTangents[3*i+2] = vertex.tangent.z();
    }
    if ((int)vertexUvs.size() >= (i+1)*2) {
        vertexUvs[2*i] = vertex.u;
        vertexUvs[2*i+1] = vertex.v;
    }
}

void QuadMesh::setQuadAt(int i, int p0, int p1, int p2, int p3)
{
    quadIndices[4*i] = p0;
    quadIndices[4*i+1] = p1;
    quadIndices[4*i+2] = p2;
    quadIndices[4*i+3] = p3;
}

int QuadMesh::getNumVertices() const
{
    return (int)vertexPositions.size()/3;
}

int QuadMesh::getNumQuads() const
{
    return (int)quadIndices.size()/4;
}

java::ArrayList<double>& QuadMesh::getVertexPositions() { return vertexPositions; }
java::ArrayList<double>& QuadMesh::getVertexNormals() { return vertexNormals; }
java::ArrayList<double>& QuadMesh::getVertexBinormals() { return vertexBinormals; }
java::ArrayList<double>& QuadMesh::getVertexTangents() { return vertexTangents; }
java::ArrayList<double>& QuadMesh::getVertexColors() { return vertexColors; }
java::ArrayList<double>& QuadMesh::getVertexUvs() { return vertexUvs; }
java::ArrayList<int>& QuadMesh::getQuadIndices() { return quadIndices; }

void QuadMesh::calculateNormals()
{
    initVertexNormalsArray();
    for (int i = 0; i < getNumQuads(); i++) {
        int i0 = quadIndices[4*i];
        int i1 = quadIndices[4*i+1];
        int i2 = quadIndices[4*i+2];
        int i3 = quadIndices[4*i+3];
        if (i0 < 0 || i1 < 0 || i2 < 0 || i3 < 0 ||
            i0 >= getNumVertices() || i1 >= getNumVertices() ||
            i2 >= getNumVertices() || i3 >= getNumVertices()) {
            continue;
        }

        Vector3Dd v0(vertexPositions[3*i0], vertexPositions[3*i0+1], vertexPositions[3*i0+2]);
        Vector3Dd v1(vertexPositions[3*i1], vertexPositions[3*i1+1], vertexPositions[3*i1+2]);
        Vector3Dd v2(vertexPositions[3*i2], vertexPositions[3*i2+1], vertexPositions[3*i2+2]);
        Vector3Dd v3(vertexPositions[3*i3], vertexPositions[3*i3+1], vertexPositions[3*i3+2]);

        Vector3Dd n1 = v1.subtract(v0).crossProduct(v2.subtract(v0)).normalized();
        Vector3Dd n2 = v2.subtract(v0).crossProduct(v3.subtract(v0)).normalized();
        Vector3Dd n = n1.add(n2).normalized();

        int ids[4] = {i0, i1, i2, i3};
        for (int k = 0; k < 4; k++) {
            int idx = ids[k];
            vertexNormals[3*idx] += n.x();
            vertexNormals[3*idx+1] += n.y();
            vertexNormals[3*idx+2] += n.z();
        }
    }

    for (int i = 0; i < getNumVertices(); i++) {
        Vector3Dd n(vertexNormals[3*i], vertexNormals[3*i+1], vertexNormals[3*i+2]);
        n = n.normalized();
        vertexNormals[3*i] = n.x();
        vertexNormals[3*i+1] = n.y();
        vertexNormals[3*i+2] = n.z();
    }
}

double* QuadMesh::calculateMinMaxPositions()
{
    double* minMax = new double[6];
    if (getNumVertices() == 0) {
        minMax[0] = minMax[1] = minMax[2] = 0.0;
        minMax[3] = minMax[4] = minMax[5] = 0.0;
        return minMax;
    }

    double minX = 1e308, minY = 1e308, minZ = 1e308;
    double maxX = -1e308, maxY = -1e308, maxZ = -1e308;

    for (int i = 0; i < getNumVertices(); i++) {
        double x = vertexPositions[3*i];
        double y = vertexPositions[3*i+1];
        double z = vertexPositions[3*i+2];
        minX = std::min(minX, x);
        minY = std::min(minY, y);
        minZ = std::min(minZ, z);
        maxX = std::max(maxX, x);
        maxY = std::max(maxY, y);
        maxZ = std::max(maxZ, z);
    }

    minMax[0] = minX; minMax[1] = minY; minMax[2] = minZ;
    minMax[3] = maxX; minMax[4] = maxY; minMax[5] = maxZ;
    return minMax;
}

double* QuadMesh::getMinMax()
{
    return calculateMinMaxPositions();
}

TriangleMeshGroup* QuadMesh::exportToTriangleMeshGroup()
{
    TriangleMesh mesh;
    mesh.setName(name);
    mesh.getVertexPositions() = vertexPositions;
    mesh.getVertexNormals() = vertexNormals;
    mesh.getVertexBinormals() = vertexBinormals;
    mesh.getVertexTangents() = vertexTangents;
    mesh.getVertexColors() = vertexColors;
    mesh.getVertexUvs() = vertexUvs;

    mesh.initTriangleArrays(getNumQuads()*2);
    java::ArrayList<int>& tri = mesh.getTriangleIndexes();

    int t = 0;
    for (int i = 0; i < getNumQuads(); i++) {
        int p0 = quadIndices[4*i];
        int p1 = quadIndices[4*i+1];
        int p2 = quadIndices[4*i+2];
        int p3 = quadIndices[4*i+3];
        tri[3*t] = p0; tri[3*t+1] = p1; tri[3*t+2] = p2; t++;
        tri[3*t] = p0; tri[3*t+1] = p2; tri[3*t+2] = p3; t++;
    }
    mesh.calculateNormals();

    TriangleMeshGroup* g = new TriangleMeshGroup();
    g->addMesh(mesh);
    return g;
}

Ray* QuadMesh::doIntersection(const Ray& inOut_Ray)
{
    TriangleMeshGroup* g = exportToTriangleMeshGroup();
    Ray* out = g->doIntersection(inOut_Ray);
    delete g;
    return out;
}

bool QuadMesh::doIntersection(const Ray& inRay, RayHit* outHit)
{
    TriangleMeshGroup* g = exportToTriangleMeshGroup();
    bool out = g->doIntersection(inRay, outHit);
    delete g;
    return out;
}

void QuadMesh::doExtraInformation(const Ray& inRay, double inT, RayHit* outData)
{
    TriangleMeshGroup* g = exportToTriangleMeshGroup();
    g->doExtraInformation(inRay, inT, outData);
    delete g;
}

java::String QuadMesh::toString() const
{
    java::String msg;
    QuadMesh* self = const_cast<QuadMesh*>(this);
    double* mm = self->getMinMax();
    msg += "- QuadMesh ------------------------------------------------------------\n";
    msg += "  - Number of quads:";
    msg += std::to_string(getNumQuads()).c_str();
    msg += "\n";
    msg += "  - Number of vertexes:";
    msg += std::to_string(getNumVertices()).c_str();
    msg += "\n";
    msg += "  - MINMAX: (";
    msg += std::to_string(mm[0]).c_str();
    msg += ", ";
    msg += std::to_string(mm[1]).c_str();
    msg += ", ";
    msg += std::to_string(mm[2]).c_str();
    msg += ") - (";
    msg += std::to_string(mm[3]).c_str();
    msg += ", ";
    msg += std::to_string(mm[4]).c_str();
    msg += ", ";
    msg += std::to_string(mm[5]).c_str();
    msg += ")\n";
    msg += "---------------------------------------------------------------------------\n";
    delete [] mm;
    return msg;
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/surface/QuadMesh.ts
import { Vector3Dd } from "../../../common/linealAlgebra/Vector3Dd.js";
import { Containment } from "../../../processing/Containment.js";
import { Ray } from "../element/Ray.js";
import { RayHit } from "../element/RayHit.js";
import { Triangle } from "../element/Triangle.js";
import { Vertex } from "../element/Vertex.js";
import { Surface } from "./Surface.js";
import { TriangleMeshGroup } from "./TriangleMeshGroup.js";

/** Indexed quad mesh. Each quad is tested as the two triangles (0,1,2) and (0,2,3). */
export class QuadMesh extends Surface<Ray, RayHit> {
    private name = "default";
    private vertexPositions: Float64Array | null = null;
    private vertexNormals: Float64Array | null = null;
    private vertexBinormals: Float64Array | null = null;
    private vertexTangents: Float64Array | null = null;
    private vertexColors: Float64Array | null = null;
    private vertexUvs: Float64Array | null = null;
    private quadIndices: Int32Array | null = null;
    private incidentQuadsPerVertexArray: number[][] | null = null;

    public getName(): string {
        return this.name;
    }
    public setName(name: string): void {
        this.name = name;
    }
    public initVertexPositionsArray(count: number): void {
        this.vertexPositions = new Float64Array(count * 3);
        this.incidentQuadsPerVertexArray = Array.from({ length: count }, () => []);
    }
    public initVertexColorsArray(): void {
        this.vertexColors = new Float64Array(this.getNumVertices() * 3);
    }
    public initVertexNormalsArray(): void {
        this.vertexNormals = new Float64Array(this.getNumVertices() * 3);
    }
    public initVertexBinormalsArray(): void {
        this.vertexBinormals = new Float64Array(this.getNumVertices() * 3);
    }
    public initVertexTangentsArray(): void {
        this.vertexTangents = new Float64Array(this.getNumVertices() * 3);
    }
    public initVertexUvsArray(): void {
        this.vertexUvs = new Float64Array(this.getNumVertices() * 2);
    }
    public initQuadArrays(count: number): void {
        this.quadIndices = new Int32Array(count * 4);
    }
    public getNumVertices(): number {
        return this.vertexPositions?.length === undefined ? 0 : this.vertexPositions.length / 3;
    }
    public getNumQuads(): number {
        return this.quadIndices?.length === undefined ? 0 : this.quadIndices.length / 4;
    }
    public getVertexPositions(): Float64Array | null {
        return this.vertexPositions;
    }
    public getVertexNormals(): Float64Array | null {
        return this.vertexNormals;
    }
    public getVertexBinormals(): Float64Array | null {
        return this.vertexBinormals;
    }
    public getVertexTangents(): Float64Array | null {
        return this.vertexTangents;
    }
    public getVertexColors(): Float64Array | null {
        return this.vertexColors;
    }
    public getVertexUvs(): Float64Array | null {
        return this.vertexUvs;
    }
    public getQuadIndices(): Int32Array | null {
        return this.quadIndices;
    }

    public setVertexes(vertexes: Vertex[]): void {
        this.initVertexPositionsArray(vertexes.length);
        this.initVertexNormalsArray();
        for (let i = 0; i < vertexes.length; i++) this.setPositionAndNormal(i, vertexes[i]!);
    }
    public getVertexAt(index: number, vertex: Vertex): void {
        this.assertVertex(index);
        const p = this.position(index);
        vertex.setPosition(p);
        if (this.vertexNormals !== null) vertex.setNormal(this.vector(this.vertexNormals, index));
        if (this.vertexBinormals !== null) vertex.setBinormal(this.vector(this.vertexBinormals, index));
        if (this.vertexTangents !== null) vertex.setTangent(this.vector(this.vertexTangents, index));
        if (this.vertexUvs !== null) {
            vertex.setU(this.vertexUvs[index * 2]!);
            vertex.setV(this.vertexUvs[index * 2 + 1]!);
        }
    }
    public setVertexAt(index: number, vertex: Vertex): void {
        this.assertVertex(index);
        this.setPositionAndNormal(index, vertex);
        if (this.vertexBinormals !== null) this.setVector(this.vertexBinormals, index, vertex.binormal);
        if (this.vertexTangents !== null) this.setVector(this.vertexTangents, index, vertex.tangent);
        if (this.vertexUvs !== null) {
            this.vertexUvs[index * 2] = vertex.u;
            this.vertexUvs[index * 2 + 1] = vertex.v;
        }
    }
    public setQuadAt(index: number, p0: number, p1: number, p2: number, p3: number): void {
        if (this.quadIndices === null || index < 0 || index >= this.getNumQuads())
            throw new RangeError("Quad index out of bounds");
        for (const point of [p0, p1, p2, p3]) this.assertVertex(point);
        this.quadIndices.set([p0, p1, p2, p3], index * 4);
        for (const point of [p0, p1, p2, p3]) this.incidentQuadsPerVertexArray?.[point]?.push(index);
    }
    public calculateNormals(): void {
        if (this.vertexPositions === null) return;
        this.initVertexNormalsArray();
        const sums = Array.from({ length: this.getNumVertices() }, () => new Vector3Dd());
        for (let q = 0; q < this.getNumQuads(); q++) {
            const [a, b, c, d] = this.quad(q);
            const n1 = this.position(b)
                .subtract(this.position(a))
                .crossProduct(this.position(c).subtract(this.position(a)));
            const n2 = this.position(c)
                .subtract(this.position(a))
                .crossProduct(this.position(d).subtract(this.position(a)));
            for (const index of [a, b, c]) sums[index] = sums[index]!.add(n1);
            for (const index of [a, c, d]) sums[index] = sums[index]!.add(n2);
        }
        for (let i = 0; i < sums.length; i++) this.setVector(this.vertexNormals!, i, sums[i]!.normalized());
    }
    public doIntersectionFirstHit(ray: Ray, out: RayHit, triangleInformation?: Int32Array | number[]): boolean {
        let nearest: { t: number; point: Vector3Dd; normal: Vector3Dd; triangle: number } | null = null;
        for (let q = 0; q < this.getNumQuads(); q++) {
            const [a, b, c, d] = this.quad(q);
            for (const [x, y, z, offset] of [
                [a, b, c, 0],
                [a, c, d, 1],
            ] as const) {
                const hit = Triangle.doIntersectionWithTriangle(
                    ray,
                    this.position(x),
                    this.position(y),
                    this.position(z),
                );
                if (hit !== null && (nearest === null || hit.getT() < nearest.t))
                    nearest = {
                        t: hit.getT(),
                        point: hit.getPoint(),
                        normal: hit.getNormal(),
                        triangle: q * 2 + offset,
                    };
            }
        }
        if (nearest === null) return false;
        out.setRay(ray.withT(nearest.t));
        out.p = nearest.point;
        out.n = nearest.normal;
        if (triangleInformation !== undefined) triangleInformation[0] = nearest.triangle;
        return true;
    }
    public override doExtraInformation(ray: Ray, t: number, out: RayHit): void {
        this.doIntersectionFirstHit(ray.withT(t), out);
    }
    public override doContainmentTest(point: Vector3Dd, tolerance: number): number {
        for (let q = 0; q < this.getNumQuads(); q++) {
            const [a, b, c, d] = this.quad(q);
            if (
                Triangle.containmentTest(this.position(a), this.position(b), this.position(c), point, tolerance) !==
                    Containment.OUTSIDE ||
                Triangle.containmentTest(this.position(a), this.position(c), this.position(d), point, tolerance) !==
                    Containment.OUTSIDE
            )
                return Containment.LIMIT;
        }
        return Containment.OUTSIDE;
    }
    public getMinMax(): Float64Array {
        const result = new Float64Array([Infinity, Infinity, Infinity, -Infinity, -Infinity, -Infinity]);
        for (let i = 0; i < this.getNumVertices(); i++) {
            const p = this.position(i);
            result[0] = Math.min(result[0]!, p.x());
            result[1] = Math.min(result[1]!, p.y());
            result[2] = Math.min(result[2]!, p.z());
            result[3] = Math.max(result[3]!, p.x());
            result[4] = Math.max(result[4]!, p.y());
            result[5] = Math.max(result[5]!, p.z());
        }
        return result;
    }
    public exportToTriangleMeshGroup(): TriangleMeshGroup {
        return new TriangleMeshGroup([this]);
    }
    public override toString(): string {
        return `- QuadMesh\n  - Number of quads: ${this.getNumQuads()}\n  - Number of vertexes: ${this.getNumVertices()}\n`;
    }
    private assertVertex(index: number): void {
        if (!Number.isInteger(index) || index < 0 || index >= this.getNumVertices())
            throw new RangeError("Vertex index out of bounds");
    }
    private position(index: number): Vector3Dd {
        this.assertVertex(index);
        return this.vector(this.vertexPositions!, index);
    }
    private vector(values: Float64Array, index: number): Vector3Dd {
        return new Vector3Dd(values[index * 3]!, values[index * 3 + 1]!, values[index * 3 + 2]!);
    }
    private setVector(values: Float64Array, index: number, vector: Vector3Dd): void {
        values[index * 3] = vector.x();
        values[index * 3 + 1] = vector.y();
        values[index * 3 + 2] = vector.z();
    }
    private setPositionAndNormal(index: number, vertex: Vertex): void {
        this.setVector(this.vertexPositions!, index, vertex.position);
        this.setVector(this.vertexNormals!, index, vertex.normal);
    }
    private quad(index: number): [number, number, number, number] {
        if (this.quadIndices === null || index < 0 || index >= this.getNumQuads())
            throw new RangeError("Quad index out of bounds");
        const i = index * 4;
        return [this.quadIndices[i]!, this.quadIndices[i + 1]!, this.quadIndices[i + 2]!, this.quadIndices[i + 3]!];
    }
}
